internal_ws/ykshim/src/test_api.rs

```diff
@@ -0,0 +1,124 @@
+use libc::size_t;
+use std::convert::TryFrom;
+use std::default::Default;
+use std::ffi::{c_void, CString};
+use std::os::raw::c_char;
+use std::ptr;
+
+use ykbh::SIRInterpreter;
+use ykcompile::{TraceCompiler, REG_POOL};
+use ykpack::{self, CguHash, Local, LocalDecl, TyIndex};
+use yktrace::sir::{self, SirTrace, SIR};
+use yktrace::tir::TirTrace;
+
+#[no_mangle]
+unsafe extern "C" fn __ykshim_sirtrace_len(sir_trace: *mut c_void) -> size_t {
+    let trace = Box::from_raw(sir_trace as *mut SirTrace);
+    let ret = trace.len();
+    Box::leak(trace);
+    ret
+}
+
+#[no_mangle]
+unsafe extern "C" fn __ykshim_tirtrace_new(sir_trace: *mut c_void) -> *mut c_void {
+    let sir_trace = Box::from_raw(sir_trace as *mut SirTrace);
+    let ret = Box::into_raw(Box::new(TirTrace::new(&SIR, &*sir_trace).unwrap())) as *mut c_void;
+    Box::leak(sir_trace);
+    ret
+}
+
+#[no_mangle]
+unsafe extern "C" fn __ykshim_tirtrace_len(tir_trace: *mut c_void) -> size_t {
+    Box::from_raw(tir_trace as *mut TirTrace).len()
+}
+
+#[no_mangle]
+unsafe extern "C" fn __ykshim_tirtrace_display(tir_trace: *mut c_void) -> *mut c_char {
+    let tt = Box::from_raw(tir_trace as *mut TirTrace);
+    let st = CString::new(format!("{}", tt)).unwrap();
+    CString::into_raw(st)
+}
+
+#[no_mangle]
+unsafe extern "C" fn __ykshim_body_ret_ty(
+    sym: *mut c_char,
+    ret_cgu: *mut CguHash,
+    ret_idx: *mut TyIndex,
+) {
+    let sym = CString::from_raw(sym);
+    let rv = usize::try_from(sir::RETURN_LOCAL.0).unwrap();
+    let tyid = SIR.body(&sym.to_str().unwrap()).unwrap().local_decls[rv].ty;
+    *ret_cgu = tyid.0;
+    *ret_idx = tyid.1;
+}
+
+#[no_mangle]
+unsafe extern "C" fn __ykshim_tracecompiler_default() -> *mut c_void {
+    let tc = Box::new(TraceCompiler::new(Default::default(), Default::default()));
+    Box::into_raw(tc) as *mut c_void
+}
+
+#[no_mangle]
+unsafe extern "C" fn __ykshim_tracecompiler_drop(comp: *mut c_void) {
+    Box::from_raw(comp as *mut TraceCompiler);
+}
+
+#[no_mangle]
+unsafe extern "C" fn __ykshim_tracecompiler_insert_decl(
+    tracecompiler: *mut c_void,
+    local: Local,
+    local_ty_cgu: CguHash,
+    local_ty_index: TyIndex,
+    referenced: bool,
+) {
+    let mut tc = Box::from_raw(tracecompiler as *mut TraceCompiler);
+    tc.local_decls.insert(
+        local,
+        LocalDecl {
+            ty: (local_ty_cgu, local_ty_index),
+            referenced,
+        },
+    );
+    Box::leak(tc);
+}
+
+/// Returns a string describing the register allocation of the specified local.
+#[no_mangle]
+unsafe extern "C" fn __ykshim_tracecompiler_local_to_location_str(
+    tc: *mut c_void,
+    local: u32,
+) -> *mut c_char {
+    let mut tc = Box::from_raw(tc as *mut TraceCompiler);
+    let rstr = format!("{:?}", tc.local_to_location(Local(local)));
+    let cstr = CString::new(rstr.as_str()).unwrap().into_raw();
+    Box::leak(tc);
```

Comment:
Instead of using `Box::from_raw`+`Box::leak` everywhere you should just do `&mut *(tc as *mut TraceCompiler)`.

Reply:
Ah, because the call to these funcs (which don't consume) is inconsequential to the lifetime of the box, right?

Reply:
Yes